File: backend/core/quantities.py
"""الكميات والوحدات — §٦.٢ — النظير الخادمي لـ packages/domain/src/quantities.ts."""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Literal

from core.money import (
    QTY_SCALE,
    DomainError,
    assert_int64,
    parse_unsigned_string,
    round_half_away_div,
)

QTY_DECIMALS = 3
DecimalPlaces = Literal[0, 1, 2, 3]
_QTY_STRING = re.compile(r"^(0|[1-9][0-9]*)(\.[0-9]{1,3})?$")


@dataclass(frozen=True)
class UnitFactor:
    """عدد وحدات المخزون الأساسية في وحدة بديلة واحدة — نسبة موجبة num/den."""

    num: int
    den: int


def parse_unit_factor(num: str, den: str = "1") -> UnitFactor:
    n = parse_unsigned_string(num)
    d = parse_unsigned_string(den)
    if n == 0 or d == 0:
        raise DomainError("invalid_unit_factor", f"{num}/{den}")
    return UnitFactor(n, d)


def parse_qty_string(value: object) -> int:
    if not isinstance(value, str):
        raise DomainError("invalid_quantity", "not a string")
    if not _QTY_STRING.fullmatch(value):
        raise DomainError("invalid_quantity", value)
    whole, _, frac = value.partition(".")
    return assert_int64(int(whole) * QTY_SCALE + int(frac.ljust(QTY_DECIMALS, "0") or "0"))


def assert_qty_precision(qty_milli: int, decimal_places: DecimalPlaces) -> int:
    step = 10 ** (QTY_DECIMALS - decimal_places)
    if qty_milli % step != 0:
        raise DomainError("invalid_precision", f"{qty_milli} @ {decimal_places}")
    return qty_milli


def format_qty_string(qty_milli: int, decimal_places: DecimalPlaces) -> str:
    assert_qty_precision(qty_milli, decimal_places)
    negative = qty_milli < 0
    abs_qty = -qty_milli if negative else qty_milli
    whole, frac_int = divmod(abs_qty, QTY_SCALE)
    frac = str(frac_int).rjust(QTY_DECIMALS, "0")[:decimal_places]
    return f"{'-' if negative else ''}{whole}{'.' + frac if decimal_places > 0 else ''}"


def to_base_qty_milli(qty_milli: int, factor: UnitFactor) -> int:
    """base = qty × num / den — يجب أن تكون دقيقة بثلاث منازل وإلا inexact_quantity."""
    assert_int64(qty_milli, "qty_milli")
    scaled = qty_milli * factor.num
    if scaled % factor.den != 0:
        raise DomainError("inexact_quantity", f"{qty_milli} × {factor.num}/{factor.den}")
    return assert_int64(scaled // factor.den, "base_qty_milli")


def from_base_qty_milli_for_display(base_milli: int, factor: UnitFactor) -> int:
    """كمية بالوحدة الأساسية → بالوحدة البديلة للعرض فقط: base × den / num مقرَّبة بالنصف بعيداً
    عن الصفر إلى ثلاث منازل (CAT-03 «أمثلة محسوبة»). لا تُستعمل في حدث أو رصيد."""
    assert_int64(base_milli, "base_qty_milli")
    return assert_int64(round_half_away_div(base_milli * factor.den, factor.num), "unit_qty_milli")

File: packages/domain/src/quantities.ts
/**
 * الكميات والوحدات — §٦.٢: مقياس الكمية ثابت عند ثلاثة منازل (أجزاء الألف)، ومعامل التحويل إلى
 * وحدة المخزون الأساسية يُحفظ على السطر بتمثيل صحيح أو نسبة صحيحة موجبة، مع قاعدة دقة تمنع
 * تقريب كميات غير قابلة للتمثيل **بصمت**.
 */
import {
  DomainError,
  QTY_SCALE,
  assertInt64,
  parseUnsignedString,
  roundHalfAwayDiv,
} from "./money";

export const QTY_DECIMALS = 3;
/** `decimal_places` للوحدة بين 0 و3 يحكم الإدخال والعرض. */
export type DecimalPlaces = 0 | 1 | 2 | 3;

/** معامل تحويل: عدد وحدات المخزون الأساسية في وحدة بديلة واحدة — نسبة موجبة num/den. */
export interface UnitFactor {
  readonly num: bigint;
  readonly den: bigint;
}

const QTY_STRING = /^(0|[1-9][0-9]*)(\.[0-9]{1,3})?$/;

/** يقبل "12" أو "12/5"؛ الصفر مرفوض في البسط والمقام. */
export function parseUnitFactor(num: string, den = "1"): UnitFactor {
  const n = parseUnsignedString(num);
  const d = parseUnsignedString(den);
  if (n === 0n || d === 0n) throw new DomainError("invalid_unit_factor", `${num}/${den}`);
  return { num: n, den: d };
}

/** كمية غير سالبة بصيغة عشرية قانونية (حتى 3 منازل) → أجزاء الألف. */
export function parseQtyString(input: unknown): bigint {
  if (typeof input !== "string") throw new DomainError("invalid_quantity", "not a string");
  if (!QTY_STRING.test(input)) throw new DomainError("invalid_quantity", input);
  const [whole, frac = ""] = input.split(".") as [string, string?];
  return assertInt64(BigInt(whole) * QTY_SCALE + BigInt((frac ?? "").padEnd(QTY_DECIMALS, "0")));
}

/** عرض الكمية بعدد المنازل المعلن للوحدة؛ الكمية يجب أن تكون قابلة للتمثيل بهذه المنازل. */
export function formatQtyString(qtyMilli: bigint, decimalPlaces: DecimalPlaces): string {
  assertQtyPrecision(qtyMilli, decimalPlaces);
  const negative = qtyMilli < 0n;
  const abs = negative ? -qtyMilli : qtyMilli;
  const whole = abs / QTY_SCALE;
  const frac = (abs % QTY_SCALE).toString().padStart(QTY_DECIMALS, "0").slice(0, decimalPlaces);
  return `${negative ? "-" : ""}${whole}${decimalPlaces > 0 ? "." + frac : ""}`;
}

/** الكمية يجب أن تكون مضاعفاً لـ10^(3 − decimal_places)؛ وإلا `invalid_precision` — لا تقريب صامت. */
export function assertQtyPrecision(qtyMilli: bigint, decimalPlaces: DecimalPlaces): bigint {
  const step = 10n ** BigInt(QTY_DECIMALS - decimalPlaces);
  if (qtyMilli % step !== 0n)
    throw new DomainError("invalid_precision", `${qtyMilli} @ ${decimalPlaces}`);
  return qtyMilli;
}

/**
 * كمية بالوحدة البديلة → كمية بوحدة المخزون الأساسية (كلاهما بأجزاء الألف):
 * base = qty × num / den — تُقبل فقط إن كانت **دقيقة** بثلاثة منازل، وإلا `inexact_quantity`.
 */
export function toBaseQtyMilli(qtyMilli: bigint, factor: UnitFactor): bigint {
  assertInt64(qtyMilli, "qty_milli");
  const scaled = qtyMilli * factor.num;
  if (scaled % factor.den !== 0n)
    throw new DomainError("inexact_quantity", `${qtyMilli} × ${factor.num}/${factor.den}`);
  return assertInt64(scaled / factor.den, "base_qty_milli");
}

/**
 * كمية بوحدة المخزون الأساسية → كمية بالوحدة البديلة **للعرض فقط** (CAT-03 «أمثلة محسوبة»):
 * unit = base × den / num مقرَّبة إلى ثلاث منازل بالنصف بعيداً عن الصفر (§٦.٢). لا تُستعمل في حدث
 * أو رصيد — الرصيد يُعرض بالوحدة الأساس؛ التقريب هنا في المجال لا في الواجهة.
 */
export function fromBaseQtyMilliForDisplay(baseMilli: bigint, factor: UnitFactor): bigint {
  assertInt64(baseMilli, "base_qty_milli");
  return assertInt64(roundHalfAwayDiv(baseMilli * factor.den, factor.num), "unit_qty_milli");
}
